File: 621-task-scheduler/task-scheduler.java
class Solution {
    // SC: O(26) -> O(1)
    // TC: O(m) where m->tasks
    public int leastInterval(char[] tasks, int n) {
        // Build freq map
        int[] freq = new int[26];

        for (char task : tasks) {
            freq[task - 'A']++;
        }

        // Convert freq to PQ
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());

        for (int cnt : freq) {
            if (cnt > 0) {
                maxHeap.add(cnt);
            }
        }
        // queue to store (count, time when available) 
        Queue<int[]> queue = new LinkedList<>();
        int time = 0;
        while (!maxHeap.isEmpty() || !queue.isEmpty()) {
            time++;
            // If heap empty, jump to queue time
            if (maxHeap.isEmpty()) {
               time = queue.peek()[1];
            }
            else {
                int count = maxHeap.poll() - 1;
                if (count > 0) {
                    queue.add(new int[]{count, time + n});
                }
            }
            if (!queue.isEmpty() && queue.peek()[1] == time) {
                maxHeap.offer(queue.poll()[0]);
            }
        }

        return time;
    }
}

File: 621-task-scheduler/task-scheduler.py
class Solution(object):
    def leastInterval(self, tasks, n):
        """
        :type tasks: List[str]
        :type n: int
        :rtype: int
        """
        # Keep count of frequencies and move them to Max PQ

        freq = Counter(tasks)
        maxHeap = [-count for count in freq.values()]
        heapq.heapify(maxHeap)

        time = 0
        # store map of -count, time
        q = deque()
          
        while maxHeap or q:
            time += 1
            if not maxHeap:
                # jump to time
                time = q[0][1] 

            else:
                count = heapq.heappop(maxHeap) + 1
                if count:
                    q.append([count, time + n])
                
            # remove from queue when time matches 
            if q and time == q[0][1]:
                heapq.heappush(maxHeap, q.popleft()[0])
            


        return time    




